cpp/include/cuspatial/experimental/detail/coordinate_transform.cuh

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cuspatial/constants.hpp>
+#include <cuspatial/types.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <thrust/transform.h>
+
+#include <iterator>
+#include <type_traits>
+
+namespace cuspatial {
+
+namespace detail {
+
+constexpr double EARTH_CIRCUMFERENCE_KM_PER_DEGREE = EARTH_CIRCUMFERENCE_EQUATOR_KM / 360.0;
+
+template <typename T>
+__device__ inline T midpoint(T a, T b)
+{
+  return (a + b) / 2;
+}
+
+template <typename T>
+__device__ inline T lon_to_x(T lon, T lat)
+{
+  return lon * EARTH_CIRCUMFERENCE_KM_PER_DEGREE * cos(lat * DEGREE_TO_RADIAN);
+};
+
+template <typename T>
+__device__ inline T lat_to_y(T lat)
+{
+  return lat * EARTH_CIRCUMFERENCE_KM_PER_DEGREE;
+};
+
+template <typename T>
+struct to_cartesian_functor {
+  to_cartesian_functor(lonlat_2d<T> origin) : _origin(origin) {}
+
+  cartesian_2d<T> __device__ operator()(lonlat_2d<T> loc)
+  {
+    return cartesian_2d<T>{lon_to_x(_origin.x - loc.x, midpoint(loc.y, _origin.y)),
+                           lat_to_y(_origin.y - loc.y)};
+  }
+
+ private:
+  lonlat_2d<T> _origin{};
+};
+
+}  // namespace detail
+
+template <class InputIt, class OutputIt, class T>
+OutputIt lonlat_to_cartesian(InputIt lon_lat_first,
+                             InputIt lon_lat_last,
+                             OutputIt xy_first,
+                             lonlat_2d<T> origin,
+                             rmm::cuda_stream_view stream)
+{
+  static_assert(std::is_floating_point_v<T>,
+                "lonlat_to_cartesian supports only floating-point coordinates.");
+
+  CUSPATIAL_EXPECTS(origin.x >= -180 && origin.x <= 180 && origin.y >= -90 && origin.y <= 90,
+                    "origin must have valid longitude [-180, 180] and latitude [-90, 90]");
+
+  return thrust::transform(rmm::exec_policy(stream),
```

Comment:
I was wondering, for methods that does not fit into the `transform` framework (e.g. linestring instance that launches on `num_points` and writes to `num_segments` output), what's the STL convention to follow? cc @jrhemstad

Reply:
What exactly at your looking for? An example of an algorithm that operates on N inputs and produces M outputs where N != M? 

Stream compaction algorithms like `copy_if` are one example ( M <= N).

Offhand I can't think of algorithms where M >= N. Maybe `set_union`? 

@codereport would be the real guru to ask here.

Reply:
For pairwise linestring distance the input the data structure is segmented, and each pair of segments (non-geometric meaning) map to a single value (distance). It's like segmented reduction with two value input iterators.

Reply:
@isVoid Do you mean two elements at a time are mapped to a single value?  Something like this:
```hs
Prelude> [1..10]                        -- [1,2,3,4,5,6,7,8,9,10]
Prelude> chunksOf 2 [1..10]             -- [[1,2],[3,4],[5,6],[7,8],[9,10]]
Prelude> map sum . chunksOf 2 $ [1..10] -- [3,7,11,15,19]
```

Reply:
@codereport no. Two separate segmented input arrays. Equal number of segments in the two arrays, but corresponding segments need not have the same length.  Apply an operator to corresponding pairs of segments.